### ISSUE TITLE
Problems w/ Implementing Crystal Graph Convolution

### ISSUE THREAD

**Issue by dong-tony** [rusty1s/pytorch_geometric#1381]
## ❓ Questions & Help

Hi!

I am doing some testing with cyrstal graph convolutions by [Xie and Grossman](https://doi.org/10.1103/PhysRevLett.120.145301) and I just want to migrate over to PyG for more flexibility instead of using the original author's code. However I am having some trouble getting the model to train (with the identical dataset that does properly train using the author's code [I am quite sure I constructed the graphs properly, but could be wrong]). I tried looked at how the weights of the linear layers within the CGConv layers are updating and noticed that they don't seem to be updating, or with very small updates, I am not sure if this is cause of the issue but it does seem abnormal. Any other suggestions for where to look for bugs? Thanks!
![image](https://user-images.githubusercontent.com/39657696/85935500-9ff32d80-b8bf-11ea-8515-3ae0a9bf4304.png)

This is how I implemented the CGConv and overall model (same architecture as the original paper)
```
class CGConv(MessagePassing):
    def __init__(self, orig_atom_fea_len, nbr_fea_len, atom_fea_len=64,
                 aggr: str = 'add', bias: bool = True, **kwargs):
        super(CGConv, self).__init__(aggr=aggr, flow='target_to_source', **kwargs)
        self.atom_fea_len = atom_fea_len
        self.nbr_fea_len = nbr_fea_len
        self.orig_atom_fea_len = orig_atom_fea_len

        self.BatchNorm1 = BatchNorm(atom_fea_len)
        self.BatchNorm2 = BatchNorm(atom_fea_len)
        self.embedding = Linear(orig_atom_fea_len, atom_fea_len)
        self.lin_f = Linear(2*atom_fea_len + nbr_fea_len, atom_fea_len, bias=bias)
        self.lin_s = Linear(2*atom_fea_len + nbr_fea_len, atom_fea_len, bias=bias)
        self.sigmoid = nn.Sigmoid()
        self.softplus1 = nn.Softplus()
        self.softplus2 = nn.Softplus() 
        
        self.reset_parameters()

    def reset_parameters(self):
        self.lin_f.reset_parameters()
        self.lin_s.reset_parameters()

    def forward(self, x: Union[Tensor, PairTensor], edge_index: Adj,
                edge_attr: OptTensor = None, size: Size = None):
        """"""
        if x.shape[1] == self.orig_atom_fea_len:
            x = self.embedding(x)
        x = self.BatchNorm1(x)
        if isinstance(x, Tensor):
            x: PairTensor = (x, x)
        # propagate_type: (x: PairTensor, edge_attr: OptTensor)
        out = self.BatchNorm2(self.propagate(edge_index, x=x, edge_attr=edge_attr, size=size))
        out = out + x[1]
        return self.softplus1(out)

    def message(self, x_i, x_j, edge_attr: OptTensor):
        z = torch.cat([x_i, x_j, edge_attr], dim=-1)
        # print(z.shape)
        return self.sigmoid(self.lin_f(z)) * self.softplus2(self.lin_s(z))

                                       
class CrystalGraphConvNet(nn.Module):

    def __init__(self, orig_atom_fea_len, nbr_fea_len,
                 atom_fea_len=64, n_conv=3, n_conv_transfer = 2, h_fea_len=128, n_h=1,
                 classification=False, transfer = False):

        super(CrystalGraphConvNet, self).__init__()
        self.classification = classification
        self.transfer = transfer
        self.convs = nn.ModuleList([CGConv(orig_atom_fea_len=orig_atom_fea_len
                                   ,atom_fea_len=atom_fea_len,
                                    nbr_fea_len=nbr_fea_len)
                                    for _ in range(n_conv)])        
        
        self.conv_to_fc_softplus = nn.Softplus()
        if n_h > 1:
            self.fcs = nn.ModuleList([nn.Linear(h_fea_len, h_fea_len)
                                      for _ in range(n_h-1)])
            self.softpluses = nn.ModuleList([nn.Softplus()
                                             for _ in range(n_h-1)])
        
        if self.classification:
            self.fc_out = nn.Linear(h_fea_len, 2)
            self.logsoftmax = nn.LogSoftmax(dim=1)
            self.dropout = nn.Dropout()  
        else:
            self.fc_out = nn.Linear(h_fea_len, 1)
            self.conv_to_fc = nn.Linear(2*atom_fea_len, h_fea_len)

    def forward(self, data):
        atom_fea, bond_index, bond_attr, batch = data.x, data.edge_index, data.edge_attr, data.batch
        for conv_func in self.convs:
            atom_fea = conv_func(x=atom_fea, edge_index=bond_index, edge_attr=bond_attr)

        crys_fea = torch.cat([gap(atom_fea, batch), gmp(atom_fea, batch)], dim=1)
        crys_fea_inter = crys_fea
        
        crys_fea = self.conv_to_fc(self.conv_to_fc_softplus(crys_fea))
        crys_fea = self.conv_to_fc_softplus(crys_fea)

        if self.classification:
            crys_fea = self.dropout(crys_fea)

        if hasattr(self, 'fcs') and hasattr(self, 'softpluses'):
            for fc, softplus in zip(self.fcs, self.softpluses):
                crys_fea = softplus(fc(crys_fea))
        
        out = self.fc_out(crys_fea)
        if self.classification:
            out = self.logsoftmax(out)
        
        return out, crys_fea_inter
```

**Comment by rusty1s** [comment]
Your implementation looks correct to me, except for that the official implementation only performs global mean pooling, see [here](https://github.com/txie-93/cgcnn/blob/master/cgcnn/model.py#L185-L186). Otherwise, it's quite hard for me to say what may go wrong here without looking at the data.

**Comment by dong-tony** [comment]
Thanks for the reply! Yeah I was playing around with the pooling method, and that isn't the thing causing the issue :(

[sample_data.zip](https://github.com/rusty1s/pytorch_geometric/files/4842218/sample_data.zip) - I included two pickled lists (or do you prefer another way for me to share?), one corresponding to the graph data I constructed and the other to the original author's data format i.e 
![image](https://user-images.githubusercontent.com/39657696/85947229-4ae30600-b917-11ea-8e3e-d37d710485ec.png)
so that original_data[i][0][0] would give the node features
![image](https://user-images.githubusercontent.com/39657696/85947243-664e1100-b917-11ea-8311-fb478e8b7ef8.png)

Let me know if you spot anything wrong with my graph data construction! Also if you need any other snippets of code, info, etc. Thanks you again!

**Comment by rusty1s** [comment]
The data is equal. However, there are many duplicated edges in there which might get filtered out in the official implementation?

**Comment by dong-tony** [comment]
Hmm, I am pretty sure the official one would just add everything up together, there isn't any further processing of the data once you input it into the model. The duplicate edges are supposed to represent the periodicity within a crystal, i.e. the edge attribute which encode the distance to the neighboring atom would be different from each parallel edge.

```
class ConvLayer(nn.Module):
    """
    Convolutional operation on graphs
    """
    def __init__(self, atom_fea_len, nbr_fea_len):
        """
        Initialize ConvLayer.

        Parameters
        ----------

        atom_fea_len: int
          Number of atom hidden features.
        nbr_fea_len: int
          Number of bond features.
        """
        super(ConvLayer, self).__init__()
        self.atom_fea_len = atom_fea_len
        self.nbr_fea_len = nbr_fea_len
        self.fc_full = nn.Linear(2*self.atom_fea_len+self.nbr_fea_len,
                                 2*self.atom_fea_len)
        self.sigmoid = nn.Sigmoid()
        self.softplus1 = nn.Softplus()
        self.bn1 = nn.BatchNorm1d(2*self.atom_fea_len)
        self.bn2 = nn.BatchNorm1d(self.atom_fea_len)
        self.softplus2 = nn.Softplus()

    def forward(self, atom_in_fea, nbr_fea, nbr_fea_idx):
        """
        Forward pass

        N: Total number of atoms in the batch
        M: Max number of neighbors

        Parameters
        ----------

        atom_in_fea: Variable(torch.Tensor) shape (N, atom_fea_len)
          Atom hidden features before convolution
        nbr_fea: Variable(torch.Tensor) shape (N, M, nbr_fea_len)
          Bond features of each atom's M neighbors
        nbr_fea_idx: torch.LongTensor shape (N, M)
          Indices of M neighbors of each atom

        Returns
        -------

        atom_out_fea: nn.Variable shape (N, atom_fea_len)
          Atom hidden features after convolution

        """
        # TODO will there be problems with the index zero padding?
        N, M = nbr_fea_idx.shape
        # convolution
        atom_nbr_fea = atom_in_fea[nbr_fea_idx, :]
        total_nbr_fea = torch.cat(
            [atom_in_fea.unsqueeze(1).expand(N, M, self.atom_fea_len),
             atom_nbr_fea, nbr_fea], dim=2)
        total_gated_fea = self.fc_full(total_nbr_fea)
        total_gated_fea = self.bn1(total_gated_fea.view(
            -1, self.atom_fea_len*2)).view(N, M, self.atom_fea_len*2)
        nbr_filter, nbr_core = total_gated_fea.chunk(2, dim=2)
        nbr_filter = self.sigmoid(nbr_filter)
        nbr_core = self.softplus1(nbr_core)
        nbr_sumed = torch.sum(nbr_filter * nbr_core, dim=1)
        nbr_sumed = self.bn2(nbr_sumed)
        out = self.softplus2(atom_in_fea + nbr_sumed)
        return out
```

**Comment by dong-tony** [comment]
I managed to solve the issue (after way too long looking at the wrong place :sweat_smile:) - everything with the convolution layer and model architecture was fine, I just wasn't careful inspecting the shape of the tensors coming out of data.y so it was messing up the loss function output and the original author's helper functions. Anyways, thanks for taking the time to help me! Greatly appreciate it.
